spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.iceberg.DeletedRowsScanTask;
+import org.apache.iceberg.ScanTaskGroup;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.spark.rdd.InputFileBlockHolder;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.JoinedRow;
+import org.apache.spark.unsafe.types.UTF8String;
+
+class ChangelogRowReader extends BaseRowReader<ChangelogScanTask> {
+  ChangelogRowReader(
+      Table table,
+      ScanTaskGroup<ChangelogScanTask> taskGroup,
+      Schema expectedSchema,
+      boolean caseSensitive) {
+    super(table, taskGroup, expectedSchema, caseSensitive);
+  }
+
+  @Override
+  protected CloseableIterator<InternalRow> open(ChangelogScanTask task) {
+    JoinedRow cdcRow = new JoinedRow();
+    cdcRow.withRight(changelogMetadata(task));
+
+    CloseableIterable<InternalRow> rows;
+    if (task instanceof AddedRowsScanTask) {
+      rows = openAddedRowsScanTask((AddedRowsScanTask) task);
+
+    } else if (task instanceof DeletedRowsScanTask) {
+      throw new UnsupportedOperationException("Deleted rows scan task is not supported yet");
+
+    } else if (task instanceof DeletedDataFileScanTask) {
+      rows = openDeletedDataFileScanTask((DeletedDataFileScanTask) task);
+
+    } else {
+      throw new IllegalArgumentException(
+          "Unsupported changelog scan task type: " + task.getClass().getName());
+    }
+
+    Preconditions.checkArgument(rows != null, "Fail to open scan task: %s", task);
+    CloseableIterable<InternalRow> cdcRows = CloseableIterable.transform(rows, cdcRow::withLeft);
+    return cdcRows.iterator();
+  }
+
+  private static InternalRow changelogMetadata(ChangelogScanTask task) {
+    InternalRow metadataRow = new GenericInternalRow(3);
+    metadataRow.update(0, UTF8String.fromString(task.operation().name()));
+    metadataRow.update(1, task.changeOrdinal());
+    metadataRow.update(2, task.commitSnapshotId());
+
+    return metadataRow;
+  }
+
+  @Override
+  protected Stream<ContentFile<?>> referencedFiles(ChangelogScanTask task) {
+    if (task instanceof AddedRowsScanTask) {
+      AddedRowsScanTask addedRowsScanTask = (AddedRowsScanTask) task;
+      DataFile file = addedRowsScanTask.file();
+      List<DeleteFile> deletes = addedRowsScanTask.deletes();
+      return Stream.concat(Stream.of(file), deletes.stream());
+
+    } else if (task instanceof DeletedRowsScanTask) {
+      throw new UnsupportedOperationException("Deleted rows scan task is not supported yet");
+
+    } else if (task instanceof DeletedDataFileScanTask) {
+      DeletedDataFileScanTask deletedDataFileScanTask = (DeletedDataFileScanTask) task;
```

Comment:
Optional: We can define helper methods that would accept specific types like for opening tasks.

```
if (task instanceof AddedRowsScanTask) {
  return addedRowsScanTaskFiles((AddedRowsScanTask) task);

} else if (task instanceof DeletedRowsScanTask) {
  throw new UnsupportedOperationException("Deleted rows scan task is not supported yet");

} else if (task instanceof DeletedDataFileScanTask) {
  return deletedDataFileScanTaskFiles((DeletedDataFileScanTask) task);

} else {
  throw new IllegalArgumentException(
      "Unsupported changelog scan task type: " + task.getClass().getName());
}
```

Reply:
Done.